drivers/video/fbdev/bcm2708_fb.c

```diff
@@ -883,87 +1034,149 @@ static int bcm2708_fb_probe(struct platform_device *dev)
 	if (!fw)
 		return -EPROBE_DEFER;
 
-	fb = kzalloc(sizeof(*fb), GFP_KERNEL);
-	if (!fb) {
-		ret = -ENOMEM;
-		goto free_region;
+	ret = rpi_firmware_property(fw,
+				    RPI_FIRMWARE_FRAMEBUFFER_GET_NUM_DISPLAYS,
+				    &num_displays, sizeof(u32));
+
+	/* If we fail to get the number of displays, or it returns 0, then
+	 * assume old firmware that doesn't have the mailbox call, so just
+	 * just set one display
+	 */
+	if (ret || num_displays == 0) {
+		num_displays = 1;
+		old = 1;
+		dev_err(&dev->dev,
+			"Unable to determine number of FB's. Assuming 1\n");
+		print_debug("Unable to determine number of FB's. Assuming 1\n");
+		ret = 0;
 	}
 
-	fb->fw = fw;
-	bcm2708_fb_debugfs_init(fb);
+	dev_info(&dev->dev, "FB found %d display(s)\n", num_displays);
 
-	fb->cb_base = dma_alloc_writecombine(&dev->dev, SZ_64K,
-					     &fb->cb_handle, GFP_KERNEL);
-	if (!fb->cb_base) {
-		dev_err(&dev->dev, "cannot allocate DMA CBs\n");
-		ret = -ENOMEM;
-		goto free_fb;
-	}
+	/* Set up the DMA information. Note we have just one set of DMA for
+	 * all the FB's, and simply ensure that the DMA only gets called
+	 * by one at a time
+	 */
+	if (!fbdev->cb_base) {
+		fbdev->cb_base = dma_alloc_writecombine(&dev->dev, SZ_64K,
+							&fbdev->cb_handle,
+							GFP_KERNEL);
+		if (!fbdev->cb_base) {
+			dev_err(&dev->dev, "cannot allocate DMA CBs\n");
+			ret = -ENOMEM;
+			goto free_fb;
+		}
 
-	pr_info("BCM2708FB: allocated DMA memory %pad\n", &fb->cb_handle);
+		dev_info(&dev->dev,
+			 "Allocated DMA memory %08x\n",
+			 fbdev->cb_handle);
 
-	ret = bcm_dma_chan_alloc(BCM_DMA_FEATURE_BULK,
-				 &fb->dma_chan_base, &fb->dma_irq);
-	if (ret < 0) {
-		dev_err(&dev->dev, "couldn't allocate a DMA channel\n");
-		goto free_cb;
+		ret = bcm_dma_chan_alloc(BCM_DMA_FEATURE_BULK,
+					 &fbdev->dma_chan_base,
+					 &fbdev->dma_irq);
+		if (ret < 0) {
+			dev_err(&dev->dev, "Couldn't allocate a DMA channel\n");
+			goto free_cb;
+		}
+		fbdev->dma_chan = ret;
+
+		ret = request_irq(fbdev->dma_irq, bcm2708_fb_dma_irq,
+				  0, "bcm2708_fb DMA", fbdev);
+		if (ret) {
+			dev_err(&dev->dev,
+				"Failed to request DMA irq\n");
+			goto free_dma_chan;
+		}
+		dev_info(&dev->dev,
+			 "Allocated DMA channel %d @ %p\n",
+			 fbdev->dma_chan, fbdev->dma_chan_base);
 	}
-	fb->dma_chan = ret;
 
-	ret = request_irq(fb->dma_irq, bcm2708_fb_dma_irq,
-			  0, "bcm2708_fb dma", fb);
-	if (ret) {
-		pr_err("%s: failed to request DMA irq\n", __func__);
-		goto free_dma_chan;
-	}
+	for (i = 0; i < num_displays && ret == 0; i++) {
+		struct bcm2708_fb *fb = &fbdev->displays[i];
 
-	pr_info("BCM2708FB: allocated DMA channel %d\n", fb->dma_chan);
+		fb->fw = fw;
```

Comment:
Why duplicate fw in each fb? Use fbdev?

Reply:
Originally because it was used in a place where the device was not available, but since I added the pointer back to the device in a later change, this can now be fixed up.